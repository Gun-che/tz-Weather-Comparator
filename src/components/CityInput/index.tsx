import React, { SyntheticEvent } from 'react'
import s from './index.module.scss'

interface ICityInputProps {
  index: number;
  onChange(e: React.ChangeEvent<HTMLInputElement & { dataset: { index: number } }>): void;
  item: [string, number];
  onClose(index: number): (e: SyntheticEvent) => void;
}

export const CityInput: React.FC<ICityInputProps> = ({
  index,
  item,
  onChange,
  onClose
}) => {

  return (
    <div className={"input-field " + s.wrap}>
      <button
        className="btn-floating btn-small waves-effect waves-light blue lighten-3"
        onClick={onClose(index)}>
        <i role="button" className="material-icons">close</i>
      </button>
      <i className="material-icons prefix">location_on</i>
      <input
        className="validate"
        data-index={index}
        type="text"
        onChange={onChange}
        id={'city' + (index + 1)}
        name={'city' + (index + 1)}
        value={item[0]}
      />
      <label htmlFor={'city' + (index + 1)}>{index + 1 + ' город'}</label>
    </div>
  )
}
