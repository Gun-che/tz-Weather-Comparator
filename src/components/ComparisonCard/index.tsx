import React from 'react'
import { IWeatherData } from '../../types/state'
import s from './index.module.scss'
import img from './wp.jpg'

export const ComparisonField: React.FC<{ data: IWeatherData[] }> = ({ data }) => {

  const sortByTemp = (): [string, number, number] => {
    let temp = -50
    let feelsLike = 0
    let city = ''
    data.forEach(i => {
      if (i.main.feels_like > temp) {
        temp = i.main.temp;
        city = i.name;
        feelsLike = i.main.feels_like
      }
    })
    return [city, temp, feelsLike]
  }

  return (
    <div className={s.wrap}>
      <div className="card">
        <div className="card-image">
          <img src={img} alt="sky" />
          <span className="card-title">{sortByTemp()[0]} cамый тёплый город, из выбранных вами!</span>
        </div>
        <div className="card-content">
          <p>Температура здесь сегодня {Math.round(sortByTemp()[1]) + `°C!`}</p>
          <p>По ощущениям {Math.round(sortByTemp()[2]) + `°C`}</p>
        </div>

      </div>
      <h2>{}</h2>
    </div>
  )
}
