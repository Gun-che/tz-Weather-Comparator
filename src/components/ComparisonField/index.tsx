import React from 'react'
import { IWeatherData } from '../../types/state'

export const ComparisonField: React.FC<{ data: IWeatherData[] }> = ({ data }) => {
  return (
    <div>

      {data.length ? <h2>{data[0].main.temp} ? {data[1].main.temp}</h2> : null}
    </div>
  )
}
