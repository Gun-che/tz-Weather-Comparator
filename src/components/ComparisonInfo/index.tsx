import React from 'react'
import { IWeatherData } from '../../types/state'

interface ISortResult {
  warmCity: string;
  coldCity: string;
  windyCity: string;
  windlessCity: string;
  wetCity: string;
  dryCity: string;
}

export const ComparisonInfo: React.FC<{ data: IWeatherData[] }> = ({ data }) => {

  const sort = (): ISortResult => {
    let warm = -100
    let cold = 100
    let windy = 0
    let windless = 100
    let wet = 0
    let dry = 100
    let warmCity = ''
    let coldCity = ''
    let windyCity = ''
    let windlessCity = ''
    let wetCity = ''
    let dryCity = ''

    data.forEach(i => {

      if (i.main.feels_like > warm) {
        warmCity = i.name;
        warm = i.main.feels_like
      }

      if (i.main.feels_like < cold) {
        coldCity = i.name;
        cold = i.main.feels_like
      }

      if (i.wind.speed > windy) {
        windyCity = i.name;
        windy = i.main.feels_like
      }

      if (i.wind.speed < windless) {
        windlessCity = i.name;
        windless = i.main.feels_like
      }

      if (i.main.humidity > wet) {
        wetCity = i.name;
        wet = i.main.feels_like
      }

      if (i.main.humidity < dry) {
        dryCity = i.name;
        dry = i.main.feels_like
      }
    })

    return {
      warmCity,
      coldCity,
      windyCity,
      windlessCity,
      wetCity,
      dryCity,
    }
  }

  const {
    warmCity,
    coldCity,
    windyCity,
    windlessCity,
    wetCity,
    dryCity, }: ISortResult = sort()

  return (
    <div>
      <div className="card blue-grey darken-1">
        <div className="card-content white-text">
          <span className="card-title">Card Title</span>
          <p>Теплее прочего: {warmCity}</p>
          <p>Холоднее прочего: {coldCity}</p>
          <p>Ветренее прочего: {windyCity}</p>
          <p>Наиенее ветренно: {windlessCity}</p>
          <p>Самая высокая влажность: {wetCity}</p>
          <p>Суше прочего: {dryCity}</p>
        </div>
      </div>
    </div>
  )
}
