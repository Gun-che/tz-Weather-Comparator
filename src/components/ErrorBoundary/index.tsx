import React from 'react'
import s from './index.module.scss'

interface IState {
  hasError: boolean;
  msg: string;
  err: number;
}

export class ErrorBoundary extends React.Component<{}, IState> {
  constructor(props: {}) {
    super(props);
    this.state = {
      hasError: false,
      msg: '',
      err: 0,
    }
  }

  static getDerivedStateFromError(error: Error) {
    return {
      hasError: true,
      msg: error.message,
      err: error.name,
    }
  }

  render() {
    const err = this.state.err;
    const msg = this.state.msg;
    const hasError = this.state.err;

    if (hasError) {
      return (
        <>
          <h1>Что-то пошлло не так, пожалуйста обновите страницу!</h1>
          <div className={s.wrap + ' container'}>
            {err && err}
            {msg && <h3>{msg}</h3>}
          </div>
        </>
      )

    } else {
      return this.props.children
    }
  }
}
