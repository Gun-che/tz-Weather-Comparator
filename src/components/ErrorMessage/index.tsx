import React from 'react'
import s from './index.module.scss'

export const ErrorMessage: React.FC<{ message: string }> = (
  { message }
) => {

  return (<>
    {message ? (<div className={s.wrap + ' containerCust'}><h2>{message}</h2></div>) : (<div className={s.emptyWrap}></div>)
    }
  </>)
}
