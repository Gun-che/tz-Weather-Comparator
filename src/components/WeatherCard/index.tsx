import React from 'react'
import { withACapitalLetter } from '../../helpers/withACapitalLetter'
import { baseIconUrl } from '../../helpers/api'
import { IWeatherData } from '../../types/apiResponse'
import { weatherMsg } from '../../helpers/messages'

export const WeatherCard: React.FC<{ data: IWeatherData }> = ({
  data: {
    name,
    visibility,
    weather: [{ id, icon, description }],
    wind: { speed },
    main: { feels_like, humidity, pressure, temp }
  }
}) => {

  const anotation = withACapitalLetter(weatherMsg[id]['ru'])

  return (

    <div className="card">
      <div className="card-image waves-effect waves-block waves-light">
        <img alt={description} className="activator" src={`${baseIconUrl}${icon}@4x.png`} />
      </div>
      <div className="card-content">
        <span className="card-title activator grey-text text-darken-4">{name}<i className="material-icons right">more_vert</i></span>
        <p>{Math.round(temp)}°C по ощущениям: {Math.round(feels_like)}°C</p>
      </div>
      <div className="card-reveal">
        <span className="card-title grey-text text-darken-4">{name}<i className="material-icons right">close</i></span>

        <p>{anotation}</p>
        <p>{Math.round(temp)}°C по ощущениям: {Math.round(feels_like)}°C</p>
        <p>Ветер: {speed}м/с</p>
        <p>Влажность: {humidity}%</p>
        <p>Давление: {pressure}гПа</p>
        <p>Видимость: {visibility}м</p>
      </div>
    </div>
  )
}
