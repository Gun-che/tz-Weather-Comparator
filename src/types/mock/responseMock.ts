import { IDailyWeatherData, IHourlyWeatherData, IWeatherData, IAllWeatherData } from "../apiResponse";

export const weatherDataMock: IWeatherData = {
  name: '1',
  visibility: 1,
  weather: [{
    id: 1,
    icon: '1',
    description: '1',
  }],
  wind: { speed: 1 },
  coord: { lon: 1, lat: 1 },
  main: {
    feels_like: 1,
    humidity: 1,
    pressure: 1,
    temp: 1,
  },
}

export const hourlyWeatherDataMock: IHourlyWeatherData = {
  clouds: 1,
  feels_like: 1,
  humidity: 1,
  pressure: 1,
  temp: 1,
  weather: [{
    id: 1,
    icon: '1',
    description: '1',
  }],
  wind_speed: 1,
  dt: 1,
}

export const dailyWeatherDataMock: IDailyWeatherData = {
  clouds: 10,
  weather: [{
    id: 1,
    icon: '1',
    description: '1',
  }],
  pressure: 10,
  feels_like: { day: 1, night: 1, eve: 1, morn: 1 },
  humidity: 1,
  sunrise: 1,
  sunset: 1,
  temp: { day: 1, night: 1, eve: 1, morn: 1 },
  wind_speed: 1,
  dt: 1,
}

export const allWeatherDataMock: IAllWeatherData = {
  hourly: [hourlyWeatherDataMock],
  daily: [dailyWeatherDataMock],
}