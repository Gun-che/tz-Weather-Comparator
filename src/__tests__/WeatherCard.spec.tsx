import React from 'react'
import { shallow } from 'enzyme'
import { WeatherCard } from '../components/WeatherCard'
import { weatherDataMock } from '../types/mock/responseMock'

describe('weatherCard component', () => {
  const props = {
    data: weatherDataMock,
  }
  describe('weatherCard component render', () => {
    const weatherCard = shallow(<WeatherCard {...props} />)

    it('renders properly', () => {
      expect(weatherCard).toMatchSnapshot()
    })

    it('render <img>', () => {
      expect(weatherCard.find('img')).toHaveLength(1)
    })

    it('render <p>', () => {
      expect(weatherCard.find('p')).toHaveLength(7)
    })
  })
})