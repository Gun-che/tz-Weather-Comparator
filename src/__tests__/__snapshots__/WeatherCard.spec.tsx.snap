// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weatherCard component weatherCard component render renders properly 1`] = `
<div
  className="card"
>
  <div
    className="card-image waves-effect waves-block waves-light"
  >
    <img
      alt="1"
      className="activator"
      src="http://openweathermap.org/img/wn/1@4x.png"
    />
  </div>
  <div
    className="card-content"
  >
    <span
      className="card-title activator grey-text text-darken-4"
    >
      1
      <i
        className="material-icons right"
      >
        more_vert
      </i>
    </span>
    <p>
      1
      °C по ощущениям: 
      1
      °C
    </p>
  </div>
  <div
    className="card-reveal"
  >
    <span
      className="card-title grey-text text-darken-4"
    >
      1
      <i
        className="material-icons right"
      >
        close
      </i>
    </span>
    <p>
      Безоблачно
    </p>
    <p>
      1
      °C по ощущениям: 
      1
      °C
    </p>
    <p>
      Ветер: 
      1
      м/с
    </p>
    <p>
      Влажность: 
      1
      %
    </p>
    <p>
      Давление: 
      1
      гПа
    </p>
    <p>
      Видимость: 
      1
      м
    </p>
  </div>
</div>
`;
